File: chapter8/ques5.cpp
/*******************************************************************************
 *  @File  : ques5.cpp
 *  @Author: Zhang Zixuan
 *  @Email : zixuan.zhang.victor@gmail.com
 *  @Blog  : www.noathinker.com
 *  @Date  : 2015年10月25日 星期日 16时02分41秒
 ******************************************************************************/

/*
 * Question:
 *
 * Implement an algorithm to print all valid (e g , properly opened and closed) combinations of n-pairs of parentheses 
 * EXAMPLE:
 * input: 3 (e g , 3 pairs of parentheses)
 * output: ()()(), ()(()), (())(), ((()))
 *
 */

#include <iostream>
#include <string>

using namespace std;

void construct_string(int deep, int num, string& str)
{
    if (num == deep)
    {
        cout<<str<<endl;
        return ;
    }
    if ("" == str)
    {
        str = "()";
        construct_string(deep+1, num, str);
        return ;
    }
    string temp = "()";
    string temp1 = temp.insert(0, str);
    construct_string(deep+1, num, temp1);
    temp.erase(0, str.length());
    string temp2 = temp.insert(1, str);
    construct_string(deep+1, num, temp2);
    temp.erase(1, str.length());
    string temp3 = temp.insert(2, str);
    if (temp3 != temp1)
        construct_string(deep+1, num, temp3);
}

int main()
{
    string str = "";
    construct_string(0, 3, str);
    return 0;
}

File: chapter8/ques5.py
"""
@Question:

    Given n pairs of parentheses, write a function to generate
    all combinations of well-formed parentheses.
    For example, given n = 3, a solution set is:
    "((()))", "(()())", "(())()", "()(())", "()()()"
"""

class Solution(object):
    def _generate(self, deep, num, string, result):
        if deep == num:
            if string not in result:
                result.append(string)
            return
        if "" == string:
            self._generate(deep+1, num, "()", result)
            return
        for i in range(0, len(string)+1):
            self._generate(deep+1, num, string[0:i]+"()"+string[i:], result)
            
    def generateParenthesis(self, n):

        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        self._generate(0, n, "", result)
        return result
        

if __name__ == "__main__":
    s = Solution()
    result = s.generateParenthesis(4)
    print result
    print len(result)
